src/mongocxx/private/collection.hpp

```diff
@@ -39,6 +39,19 @@ class collection::impl {
           client_impl(client) {
     }
 
+    impl(const impl& i) : 
+      collection_t{libmongoc::collection_copy(i.collection_t)},
+      database_name{i.database_name},
+      client_impl{i.client_impl}
+    {}
+
+    impl& operator=(const impl& i) { 
+        collection_t = libmongoc::collection_copy(i.collection_t);
```

Comment:
We might just want to make the `collection_t` member of `collection::impl` a `mongoc_collection_t* const` so that we can only copy construct but not copy assign impls, and then just keep the rest of the code the same.